pkg/actions/lua/lakefs/catalogexport/symlink_exporter.lua

```diff
@@ -0,0 +1,93 @@
+local extractor = require("lakefs/catalogexport/table_extractor")
+local hive = require("lakefs/catalogexport/hive")
+local utils = require("lakefs/catalogexport/internal")
+local url = require("net/url")
+local pathlib = require("path")
+local strings = require("strings")
+
+-- make it a function so that @nopcoder can ask why we bloat the code? 
+local function parse_storage_uri(uri)
+    local parsed_uri = url.parse(uri)
+    local storage = parsed_uri.scheme
+    local bucket = parsed_uri.host
+    local key = parsed_uri.path
+    if key ~= nil then
+        key = key:sub(2) -- remove the leading slash
+    end
+    return storage, bucket, key
+end
+
+--[[
+    ### Default Symlink File(s) structure:
+    
+    ${storageNamespace}
+    _lakefs/
+        exported/
+            ${ref}/
+                ${commitId}/
+                    ${tableName}/
+                        p1=v1/symlink.txt
+                        p1=v2/symlink.txt
+                        p1=v3/symlink.txt
+                        ...
+]]
+local function gen_storage_uri_prefix(storage_ns, commit_id, action_info)
+    local branch_or_tag = utils.ref_from_branch_or_tag(action_info)
+    return pathlib.join("/", storage_ns, "_lakefs/exported", branch_or_tag, utils.short_digest(commit_id), "/")
+end
+
+local function new_std_table_writer(export_base_uri)
+    local storage, bucket, key_prefix = parse_storage_uri(export_base_uri)
+    return function(suffix_key, data)
+        local key = pathlib.join("/", key_prefix, suffix_key)
+        print("[STD Writer Put Object] Bucket " .. bucket .. " key: " .. key .. "  content: \n" .. data)
+    end
+end
+
+local function new_s3_table_writer(client, export_base_uri)
+    local storage, bucket, key_prefix = parse_storage_uri(export_base_uri)
+    return function(suffix_key, data)
+        local key = pathlib.join("/", key_prefix, suffix_key)
+        client.put_object(bucket, key, data)
+        print("S3 Put: bucket: " .. bucket .. " key: " .. key)
+    end
+end
+
+local function export(lakefs_client, repo_id, commit_id, table_src_path, blockstore_write, trim_obj_base_path)
```

Comment:
The `export` accepts a lot of arguments and one of them is a specific callback that do two things:
1. format the target path (which is the same code for both callback)
2. write the information (s3 or stdout)
Suggest to turn `export` to iterator so you can implement `export_s3` as simple loop which grab the path information and write it as the callback (just without a callback) to the s3 bucket (same for every target)
3. I don't think that export should handle argument like `trim_object_base_path` - the caller doesn't know what to pass. Another reason not to expose this function and let the user call export_s3 which knows how to handle the path.
For conditional format of the data in the export output - suggest passing a table argument with optional members and one of them can be this flag, or better, as iterator it can return a table of the parts and the caller s3 or std can format as they should the information.

Reply:
changed it as we discussed!